macros/support/src/class/class.rs

```diff
@@ -29,12 +27,85 @@ pub struct InterfaceMethod {
+        let struct_tokens = self.to_struct_tokens();
+        let class_trait_impl_tokens = self.to_class_trait_impl_tokens();
+        let class_factory = super::class_factory::generate(self);
+        let vtable_static_items = self.gen_vtable_static_items();
+        let interface_from_impls = self.interface_from_impls();
+
+        quote! {
+            #struct_tokens
+            #class_trait_impl_tokens
+            #vtable_static_items
+            #interface_from_impls
+            #class_factory
+        }
+    }
+
+    /// Creates static items containing the vtables for each top-level interface.
+    fn gen_vtable_static_items(&self) -> TokenStream {
+        self.interfaces
+            .iter()
+            .enumerate()
+            .map(move |(index,  interface)| {
+                let interface_name = &interface.path;
+                let interface_tokens = interface.to_initialized_vtable_tokens(self, index);
+                let vtable_item_ident = interface.vtable_static_item_ident(self);
+                quote! {
+                    #[allow(non_upper_case_globals)]
+                    static #vtable_item_ident: <#interface_name as ::com::Interface>::VTable = #interface_tokens;
+                }
+            }).collect()
+    }
 
-        TokenStream::from_iter(out)
+    /// Generates `From` impls for the interfaces implemented by this class.
+    ///
+    /// Since we know which interfaces this class implements, we can generate
+    /// `From` impls that convert directly to those interfaces. These allow apps
+    /// to obtain an interface pointer without a fallible conversion ending in
+    /// in `server.unwrap()`.
+    ///
+    /// Classes may implement one or more interface chains, where each chain
+    /// is a sequence of single-inheritance relationships. Two chains may
+    /// contain duplicates, that is, may derive from a shared subclass. When we
+    /// generate `From` impls, we have to avoid generating duplicate definitions
+    /// for these interfaces. To do so, we use a `HashSet` to track which
+    /// interfaces we have already processed.
+    ///
+    /// Because our traversal is dependent on the order of declarations that
+    /// were provided by the user, our output is deterministic, and is not
+    /// affected by the iteration order of `HashSet` (which is not
+    /// deterministic).
+    fn interface_from_impls(&self) -> TokenStream {
+        let mut interfaces_seen = std::collections::HashSet::<&syn::Path>::new();
+        let mut output = TokenStream::new();
+
+        for (index, interface) in self.interfaces.iter().enumerate() {
+            let class_name = &self.name;
+            let chain_ident = interface.chain_ident(index);
+            let ref_count_ident = crate::utils::ref_count_ident();
+
+            for interface_path in interface.iter_chain() {
+                // Avoid generating duplicate From implementations
+                if interfaces_seen.contains(interface_path) {
```

Comment:
Nit you can do `.filter(|interface_path| interfaces_seen.contains(interface_path)` on the iterator above.

Reply:
That wouldn't work because of the shared capture of `interfaces_seen`. The loop body modifies `interfaces_seen`, so it needs mutable access.

Reply:
Alternatively you can write `if !interfaces_seen.insert(interface_path) {`. `fn insert()` already returns whether the insertion succeeded, hence if the item was previously in the set or not. Should save an extra lookup.

At that point you can write a super-simple `.filter(|interface_path| interfaces_seen.insert(interface_path))` to deduplicate.